### DISCUSSION TITLE
Brainstorm: refactor to support registering custom `ValueNode` subclasses?

### DISCUSSION THREAD

**Discussion by Jasha10** [omry/omegaconf#874] [Ideas]
#### Idea:

I wonder if OmegaConf could be refactored so that adding support for a new, custom node type would be as simple as calling e.g. `OmegaConf.register_custom_node_type(MyPathNode, pathlib.Path)`. This would be a step in the direction of enabling 3rd party packages to provide support for custom node types (`pip install omegaconf-pathlib`).

#### Background:

OmegaConf implements support for a core set of data types. For example:
- [`StringNode`](https://github.com/omry/omegaconf/blob/5e73ee6a1c0f3a63fc420bf856750b80001a5644/omegaconf/nodes.py#L148) handles data typed as `str`
- [`IntegerNode`](https://github.com/omry/omegaconf/blob/5e73ee6a1c0f3a63fc420bf856750b80001a5644/omegaconf/nodes.py#L182) handles data typed as `int`
- [`AnyNode`](https://github.com/omry/omegaconf/blob/5e73ee6a1c0f3a63fc420bf856750b80001a5644/omegaconf/nodes.py#L112) handles untyped leaf-node data (and data typed as `Any`)
- etc.

We have several open feature requests and PRs looking to expand the set of supported node types. Here is a (possibly-incomplete) list:
- Support for `bytes` (#844 and #872)
- Support for `pathlib.Path` (#97 and #873)
- Support for `datetime.datetime` (#851)
- Support for `typing.Literal` (#422 and #865)
- Support for `typing.Union` (#144)
- Support for `tuple` (#392)

While supporting common use-cases is a goal of OmegaConf, there are also risks involved with taking on many new node types:
- maintenance burden
- feature bloat

Some of the above node types are easier to implement than others. In the simplest case, support for a new leaf node type (such e.g. `pathlib.Path`) requires changing OmegaConf's source code in two places:
- define a new subclass of [`ValueNode`](https://github.com/omry/omegaconf/blob/5e73ee6a1c0f3a63fc420bf856750b80001a5644/omegaconf/nodes.py#L20) must be defined (e.g. `class PathNode(ValueNode): ...`).
- Extend the switch statement in [`_node_wrap`](https://github.com/omry/omegaconf/blob/5e73ee6a1c0f3a63fc420bf856750b80001a5644/omegaconf/omegaconf.py#L1000) to handle the new type.

A simple case such as this could be handled by a `OmegaConf.register_custom_node_type` method, enabling extension as described in the introduction above.

**Comment by omry** [comment]
Hi @Jasha10,
I found your note when I was designing exactly such a system and was looking for the original pathlib.Path PR :).
I will post a design doc for review soon.

**Comment by omry** [comment]
RFC on new type adapter design for OmegaConf:

https://github.com/omry/omegaconf/issues/1288
